activeTab extension has access to page after turning the extension off, then back on

b-weinstein opened the pull request:
#### 7b1151eb72481d791968bce427be4775ff616789
<pre>
activeTab extension has access to page after turning the extension off, then back on
<a href="https://bugs.webkit.org/show_bug.cgi?id=270454">https://bugs.webkit.org/show_bug.cgi?id=270454</a>
<a href="https://rdar.apple.com/124016968">rdar://124016968</a>

Reviewed by Timothy Hatcher.

When an extension is unloaded, clear the map of open tabs and windows. That will clear any active tab permissions,
and lead to new window and tab IDs getting created if the extension is turned back on.

While we&apos;re here, clear a bunch more extension state that should be refreshed if the extension is loaded again.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload):

Canonical link: <a href="https://commits.webkit.org/275651@main">https://commits.webkit.org/275651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c810b0f256d32ef362a982546c0a535ebaeb058a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21448 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44824 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45030 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38547 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18796 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43004 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18372 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46498 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37902 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17246 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/14201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/18865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9488 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18510 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -300,6 +300,9 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     m_contentScriptWorld = nullptr;
 
+    m_tabMap.clear();
+    m_windowMap.clear();
```
m_windowOrderVector, m_focusedWindowIdentifier, m_extensionPageTabMap, m_actionWindowMap, m_actionTabMap, m_defaultAction, m_popupPageActionMap, should be cleared too.

m_ports, m_portQueuedMessages, m_nativePortMap for completeness (though unrelated to windows and tabs)

b-weinstein replied:
I'll clear these.

b-weinstein committed:
activeTab extension has access to page after turning the extension off, then back on
https://bugs.webkit.org/show_bug.cgi?id=270454
rdar://124016968

Reviewed by Timothy Hatcher.

When an extension is unloaded, clear the map of open tabs and windows. That will clear any active tab permissions,
and lead to new window and tab IDs getting created if the extension is turned back on.

While we're here, clear a bunch more extension state that should be refreshed if the extension is loaded again.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload):

Canonical link: https://commits.webkit.org/275651@main
